flow/hooks/hooks.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2018 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Operation hooks."""
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class _HooksList(list):
+    def __call__(self, *args, **kwargs):
+        """Call all hook functions as part of this list."""
+        for hook in self:
+            logger.debug(f"Executing hook function '{hook}'.")
+            try:
+                hook(*args, **kwargs)
+            except Exception as error:
+                logger.error(
+                    "Error occurred during execution of "
+                    "hook '{}': {}.".format(hook, error)
+                )
+                raise
+
+
+class Hooks:
+    """:class:`~.Hooks` execute an action or set of actions at specific stages of operation execution.
+
+    :class:`~.Hooks` can execute a user defined function when an operation
+    starts, succeeds, fails, or finishes (regardless of whether the operation
+    executed successfully or failed).
+
+    Hooks can be installed at the operation level as decorators,
+    or at the :class:`~.FlowProject` level.
+
+    Examples
+    --------
+    The example below shows an operation level decorator that prints the
+    operation name and job id at the start of the operation execution.
+
+    .. code-block:: python
+        @FlowProject.operation
+        @FlowProject.hook.on_start(lambda operation_name, job: print(
+            f"Starting operation {operation_name} on job {job.id}"))
+        def op(job):
+            pass
```

Comment:
This example here is no longer correct and we should probably also add an example for project-level hooks.

Reply:
I fixed this example but realized there are a few missing pieces. I'll write a separate comment.